src/librustc_mir/transform/cs_propagate.rs

```diff
@@ -0,0 +1,425 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! This is Constant-Simplify propagation pass. This is a composition of three distinct
+//! dataflow passes: alias-propagation, constant-propagation and terminator simplification.
+//!
+//! All these are very similar in their nature:
+//!
+//!                  | Constant  | Simplify  |
+//! |----------------|-----------|-----------|
+//! | Lattice Domain | Lvalue    | Lvalue    |
+//! | Lattice Value  | Constant  | Constant  |
+//! | Transfer       | x = const | x = const |
+//! | Rewrite        | x → const | T(x) → T' |
+//! | Bottom         | {}        | {}        |
+//! | Join           | intersect | intersect |
+//!
+//! For all of them we will be using a lattice of `HashMap<Lvalue, Either<Lvalue, Constant, Top>>`.
+
+use rustc_data_structures::fnv::FnvHashMap;
+use rustc::middle::const_val::ConstVal;
+use rustc::mir::repr::*;
+use rustc::mir::tcx::binop_ty;
+use rustc::mir::transform::{Pass, MirPass, MirSource};
+use rustc::mir::visit::{MutVisitor};
+use rustc::ty::TyCtxt;
+use rustc_const_eval::{eval_const_binop, eval_const_unop, cast_const};
+use std::collections::hash_map::Entry;
+
+use super::dataflow::*;
+
+pub struct CsPropagate;
+
+impl Pass for CsPropagate {}
```

Comment:
> I don't like highly visible names as acronyms.

Highly… visible?

> Why is the 's' lower case?

Because of how naming conventions in Rust work.